beacon-chain/sync/data_column_sidecars.go

```diff
@@ -39,249 +39,394 @@ type DataColumnSidecarsParams struct {
-//     from connected peers that are expected to custody the required data.
+// For each block in `roBlocks` that has commitments, the function attempts to obtain
+// all sidecars corresponding to the indices listed in `requestedIndices`.
 //
-// The function returns a map of block roots to their corresponding verified read-only data
-// columns. It returns an error if data column storage is unavailable, if storage/reconstruction
-// operations fail unexpectedly, or if not all requested columns could be retrieved from peers.
+// The function returns:
+//   - A map from block root to the sidecars successfully retrieved.
+//   - A set of block roots for which not all requested sidecars could be retrieved.
+//
+// Retrieval strategy (proceeds to the next step only if not all requested sidecars
+// were successfully obtained at the current step):
+//  1. Attempt to load the requested sidecars from storage, reconstructing them from
+//     other available sidecars in storage if necessary.
+//  2. Request any missing sidecars from peers. If some are still missing, attempt to
+//     reconstruct them using both stored sidecars and those retrieved from peers.
+//  3. Request all remaining possible sidecars from peers that are not already in storage
+//     or retrieved in step 2. Stop once either all requested sidecars are retrieved,
+//     or enough sidecars are available (from storage, step 2, and step 3) to reconstruct
+//     the requested ones.
 func FetchDataColumnSidecars(
 	params DataColumnSidecarsParams,
 	roBlocks []blocks.ROBlock,
-	indicesMap map[uint64]bool,
-) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, error) {
-	if len(roBlocks) == 0 || len(indicesMap) == 0 {
-		return nil, nil
+	requestedIndices map[uint64]bool,
+) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, map[[fieldparams.RootLength]byte]map[uint64]bool, error) {
+	if len(roBlocks) == 0 || len(requestedIndices) == 0 {
+		return nil, nil, nil
 	}
 
-	indices := sortedSliceFromMap(indicesMap)
-	slotsWithCommitments := make(map[primitives.Slot]bool)
-	missingIndicesByRoot := make(map[[fieldparams.RootLength]byte]map[uint64]bool)
-	indicesByRootStored := make(map[[fieldparams.RootLength]byte]map[uint64]bool)
-	result := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn)
+	// Extract slots and roots corresponding to blocks with commitments.
+	slotsWithCommitments, storedIndicesByRoot, incompleteRoots, err := commitmentsInfo(params.Storage, roBlocks)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "commitments info")
+	}
 
-	for _, roBlock := range roBlocks {
-		// Filter out blocks without commitments.
-		block := roBlock.Block()
-		commitments, err := block.Body().BlobKzgCommitments()
-		if err != nil {
-			return nil, errors.Wrapf(err, "get blob kzg commitments for block root %#x", roBlock.Root())
-		}
-		if len(commitments) == 0 {
-			continue
-		}
+	initialMissingRootCount := len(incompleteRoots)
```

Comment:
nit: move this to where it's used before the log part after requestSidecarsFromStorage